compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -1593,45 +1599,124 @@ private BObjectType getDistinctObjectType(BLangTypeDefinition typeDefinition, BO
         return definedObjType;
     }
 
-    private BType getCyclicDefinedType(BLangTypeDefinition typeDef, SymbolEnv env) {
-        BUnionType unionType = BUnionType.create(null, new LinkedHashSet<>());
-        unionType.isCyclic = true;
-        Name typeDefName = names.fromIdNode(typeDef.name);
-
-        BTypeSymbol typeDefSymbol = Symbols.createTypeSymbol(SymTag.UNION_TYPE, Flags.asMask(typeDef.flagSet),
-                typeDefName, env.enclPkg.symbol.pkgID, unionType, env.scope.owner,
-                typeDef.name.pos, SOURCE);
-
-        typeDef.symbol = typeDefSymbol;
-        unionType.tsymbol = typeDefSymbol;
-
-        // We define the unionType in the main scope here
+    private void defineTypeInMainScope(BTypeSymbol typeDefSymbol, BLangTypeDefinition typeDef) {
         if (PackageID.isLangLibPackageID(this.env.enclPkg.packageID)) {
             typeDefSymbol.origin = BUILTIN;
             handleLangLibTypes(typeDef);
         } else {
             defineSymbol(typeDef.name.pos, typeDefSymbol);
         }
+    }
+
+    private BType defineSymbolsForCyclicTypeDefinitions(BLangTypeDefinition typeDef, SymbolEnv env) {
+        Name newTypeDefName = names.fromIdNode(typeDef.name);
+        BTypeSymbol typeDefSymbol;
+        BType newTypeNode;
+
+        switch (typeDef.typeNode.getKind()) {
+            case TUPLE_TYPE_NODE:
+                newTypeNode = new BTupleType(null, new ArrayList<>());
+                typeDefSymbol = Symbols.createTypeSymbol(SymTag.TUPLE_TYPE, Flags.asMask(typeDef.flagSet),
+                        newTypeDefName, env.enclPkg.symbol.pkgID, newTypeNode, env.scope.owner,
+                        typeDef.name.pos, SOURCE);
+                break;
+            default:
+                newTypeNode = BUnionType.create(null, new LinkedHashSet<>());
```

Comment:
Do we need to create a new type every time. Don't we have the previously created type already in the `foundSym`?